## 读取数据

In [28]:
import pandas as pd
movies_df = pd.read_json('movie300new_mobile_info.json')
choosed_columns = ['actors',        #参演人员，所有的演员的信息    ==待处理==
                   'aka',           #别名
                   'card_subtitle', #日期，国家，类型，主演....
                   'comment_count', #评论数量
                   'countries',     #国家 ， 好像card_subtitle里面有，暂时保留
                   'cover.author.avatar', #编辑这个的编辑的头像
                   'cover.author.id', #编辑这个电影的编辑的id
                   'cover.author.loc.name', #编辑这个电影的编辑所在地区
                   'cover.author.loc.uid',#编辑这个电影的编辑所在地区的id（拼音）
                   'cover.author.name', ##编辑这个电影的编辑的姓名
                   'cover.author.uid', #编辑这个电影的编辑的id
                   'cover.author.url', #编辑这个电影的编辑的个人主页
                   'cover.comments_count', #编辑这个电影的Cover的评论数
                   'cover.create_time',    #编辑这个电影的Cover创建时间
                   'directors',      #导演信息    ==待处理==
                   'durations',      #电影的时长
                   'genres',         #电影的类型
                   'has_linewatch',  #有没有在线观看的链接
                   'id',             #影片的id，鉴别不同的影片
                   'info_url',       #url链接里面包含了：片名，原名，又名，导演，主演，上映，类型，片场，地区，语言，IMDB的链接
                   'intro',          #电影简介
                   'is_douban_intro', #是不是豆瓣的简介，还是其它地方的
                   'languages',      #语言类型
                   'lineticket_url', #可以在线购买的链接（和猫眼合作的）
                   'original_title', #原名，里面有部分是空值
                   'pubdate',        #公映时间和发布地方
                   'rating.count',   #打分的数量
                   'rating.star_count', #以5星为满星的 分值
                   'rating.value',   #以10分为满分的 分值   ===？？和上一个不同，是不是有专业用户和一般用户的综合？或者中和imdb？？？
                   'release_date',   #和前面的公映时间pubdate一样，少了发布的地方
                   'review_count',   #暂时不知道是什么意思，留着
                   'title',          #影片名
                   'trailer.create_time', #试看的时间，应该可以和pubtime做一个对比，看一般多久会出试看视频
                   'trailer.n_comments', #预告片的评论数，感觉应该和影片评论数正相关
                   'trailer.runtime',    #预告片的时长
                   'year'            #年，==？？暂时不知道什么时间
                  ]
movie300_choosed_column = movies_df[choosed_columns]
#actor中，准备只提取一号演员的id姓名和角色，主演中的主演
#写一个函数:
def get_actors_info(actor):
    try:
        actor_id = actor[0]['id']
        actor_name = actor[0]['name']
        actor_roles = actor[0]['roles']
    except:
        actor_id = ''
        actor_name = ''
        actor_roles = ''
    return actor_id,actor_name,actor_roles
movie300_choosed_column['actor_id'],movie300_choosed_column['actor'],movie300_choosed_column['actor_roles'] = zip(*list(movie300_choosed_column['actors'].apply(lambda x:get_actors_info(x))))
#下面写个函数，提出导演的信息，发现可以用和演员一样的玩儿法
movie300_choosed_column['director_id'],movie300_choosed_column['director'],movie300_choosed_column['director_roles'] = zip(*list(movie300_choosed_column['directors'].apply(lambda x:get_actors_info(x))))
drop_list = ['directors','actors']
movie300_choosed_column = movie300_choosed_column.drop(drop_list,axis=1)
movie300_choosed_column.sample(5)

/home/bigdata/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:53: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
/home/bigdata/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:55: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy


aka  \
114                               [心跳纽约(港), 纽约寂寞男孩(台)]   
63   [淑女鸟(台), 不得鸟小姐(港), 大鸟萌妹(豆友译名), 伯德夫人, 柏德小姐, 小仙鸟...   
22   [宠物小精灵：就决定是你了, 神奇宝贝：就决定是你了, Pokémon the Movie:...   
120     [阴阳路之常在你左右, Accompany You, Always Be with You]   
93             [我想吃了你的胰脏, I Want to Eat Your Pancreas]   

                                       card_subtitle  comment_count  \
114          2017 / 美国 / 剧情 / 马克·韦布 / 卡勒姆·特纳 杰夫·布里吉斯            396   
63   2017 / 美国 / 剧情 喜剧 家庭 / 格蕾塔·葛韦格 / 西尔莎·罗南 劳里·梅特卡夫          23830   
22           2017 / 日本 / 动画 奇幻 冒险 / 汤山邦彦 / 松本梨香 大谷育江           1025   
120           2017 / 香港 中国大陆 / 犯罪 悬疑 / 邱礼涛 / 古天乐 张智霖           2983   
93               2017 / 日本 / 剧情 爱情 / 月川翔 / 滨边美波 北村匠海           1624   

      countries                                cover.author.avatar  \
114        [美国]  https://img3.doubanio.com/icon/up139015254-13.jpg   
63         [美国]    https://img3.doubanio.com/icon/up52071601-3.jpg   
22         [日本]   https://img3.doubanio.com/icon/up3273352-362.jpg   
120  [香港, 中国大陆]   https://img3.doubanio.com/icon/up146443262-4.jpg   
93         [日本]    https://img3.doubanio.com/icon/up53294996-4.jpg   

     cover.author.id cover.author.loc.name cover.author.loc.uid  \
114        139015254                    洛阳              luoyang   
63          52071601                    武汉                wuhan   
22           3273352                    广州            guangzhou   
120        146443262                    北京              beijing   
93          53294996                    南京              nanjing   

    cover.author.name cover.author.uid          ...           \
114             横道世之介        139015254          ...            
63           Torrilla         52071601          ...            
22                  朗     ChanKingLong          ...            
120               你看看        146443262          ...            
93               红色鲱鱼         53294996          ...            

    trailer.create_time  trailer.n_comments trailer.runtime  year actor_id  \
114          2017-08-17                 0.0           02:25  2017  1328007   
63           2017-12-06                31.0           02:29  2017  1022004   
22           2017-06-14                 5.0           01:32  2017  1040476   
120          2017-09-22                26.0           01:32  2017  1027577   
93           2017-09-16                 4.0           01:32  2017  1324483   

      actor   actor_roles director_id director        director_roles  
114  卡勒姆·特纳          [演员]     1049688    马克·韦布          [导演, 制片, 编剧]  
63   西尔莎·罗南          [演员]     1022652  格蕾塔·葛韦格  [演员, 编剧, 配音, 导演, 制片]  
22     松本梨香      [演员, 导演]     1029060     汤山邦彦              [导演, 编剧]  
120     古天乐  [演员, 编剧, 制片]     1274313      邱礼涛  [导演, 摄影, 编剧, 演员, 制片]  
93     滨边美波          [演员]     1372453      月川翔          [导演, 编剧, 剪辑]  

[5 rows x 40 columns]

In [29]:
comments = pd.read_json('structure_comments.json')
comments.sample(5)

comment         create_time  \
81334                       咋欧洲人也有碟仙一说？看来这种灵异游戏不是亚洲的专利啊！ 2018-01-30 23:35:49   
23410                                      剧本真的烂 很多矛盾的地方 2017-11-14 13:02:58   
73961  李钟硕太烂了，他这个角色的呈现让吴亦凡上都可以搞定。我最喜欢金明民，以及帮助他开了两次逮捕令... 2017-09-30 01:26:55   
57711                          虽然情节单薄低幼，但是那个微信气泡狗真是把我萌翻了 2017-11-06 00:26:58   
62594                一部悬疑片引发出对女权的思考，描述了当时英国大环境下的女性地位。很惊喜 2017-10-25 04:24:57   

               id  index  rating.value  \
81334  1315833480    166           3.0   
23410  1271020215    492           3.0   
73961  1188644661     70           4.0   
57711  1267204970    160           3.0   
62594  1261617554    456           4.0   

                                             sharing_url      title  \
81334  https://www.douban.com/doubanapp/dispatch?uri=...         灵蚀   
23410  https://www.douban.com/doubanapp/dispatch?uri=...  24小时：末路重生   
73961  https://www.douban.com/doubanapp/dispatch?uri=...      杀人优越权   
57711  https://www.douban.com/doubanapp/dispatch?uri=...     表情奇幻冒险   
62594  https://www.douban.com/doubanapp/dispatch?uri=...   莱姆豪斯的杀人魔   

                                              url  \
81334  https://movie.douban.com/subject/27126455/   
23410  https://movie.douban.com/subject/26724807/   
73961  https://movie.douban.com/subject/26858970/   
57711  https://movie.douban.com/subject/26577354/   
62594  https://movie.douban.com/subject/26616731/   

                                            user.avatar    user.id  \
81334    https://img3.doubanio.com/icon/u97839460-1.jpg   97839460   
23410  https://img3.doubanio.com/icon/up163710765-2.jpg  163710765   
73961   https://img3.doubanio.com/icon/up3594852-12.jpg    3594852   
57711   https://img3.doubanio.com/icon/up2430206-15.jpg    2430206   
62594   https://img3.doubanio.com/icon/up53556694-5.jpg   53556694   

      user.loc.name user.loc.uid        user.name    user.uid  \
81334          None         None            铃兰的冬天    97839460   
23410          None         None           在在在在看你   163710765   
73961          None         None              葛团妹     3594852   
57711            北京      beijing  Fujihime Arashi  fengzaifei   
62594            南京      nanjing          胶原蛋白小天使    53556694   

                                       user.url  vote_count  
81334   https://www.douban.com/people/97839460/           0  
23410  https://www.douban.com/people/163710765/           0  
73961    https://www.douban.com/people/3594852/           0  
57711    https://www.douban.com/people/2430206/           0  
62594   https://www.douban.com/people/53556694/           0

In [30]:
#感觉电影中的现有评分对大部分人有影响，这个是属于羊群效应，所以把现有评分加入到评分做一个feature
#先学会怎么做NLP再做这个工作

In [31]:
comments = pd.merge(comments,movie300_choosed_column[['title','rating.value']],how='left',on='title')

### 采用评论的内容和打分进行分析
* 小于等于2是差评
* 等于3是中评
* 大于等于4是好评

#### 重新修正为2分问题
* 小于等于3是差评
* 大于等于4是好评

In [32]:
npl_df = comments[['comment','title','rating.value_y','rating.value_x']]

In [33]:
len(npl_df)

96596

In [34]:
npl_df.describe()

rating.value_y  rating.value_x
count    96359.000000    93655.000000
mean         6.770185        3.350456
std          0.942223        1.003897
min          5.000000        1.000000
25%          6.100000        3.000000
50%          6.700000        3.000000
75%          7.400000        4.000000
max          9.200000        5.000000

去掉npl_df中的nan值

In [35]:
npl_df = npl_df.dropna()

In [36]:
npl_df.sample(10)

comment     title  \
13751                            首先，我是看到王敏奕主演才看的，然后电影还不错      初恋日记   
34791  最后结局解释了子弹的bug，但“过去的故事”还是无法与之前的相比，本作是又一波屠戮开始的前奏...  电锯惊魂8：竖锯   
34701  人类再次回想起被哥斯拉胖揍的恐怖，结尾好评不过也猜得到，如果只是无脑的打打打应该凑不了三部的...  哥斯拉：怪兽行星   
95405                                   二星为了这电影拍摄环境艰苦给的。      远山恋人   
68093                                   三流动作片 故事情节处理也是不行      赶尽杀绝   
79302                                           个人评分7.3分     你好布拉德   
40023            a poor masturbation but still orgasmic.        酒会   
88936  整体就是故作玄虚的做作 但看到最后竟然觉得有点惊喜 感觉导演想要在最后来个主题升华 而前面的...       方法派   
40999  剧情其实有一点勉强 但是这个颜值是让人无法拒绝啊 女主男主女二男二全都好看呐 翔平alan真...   昼行闪耀的流星   
56672                                           很细腻，对胃口。    时间中的孩子   

       rating.value_y  rating.value_x  
13751             5.6             4.0  
34791             6.9             2.0  
34701             6.3             4.0  
95405             6.0             2.0  
68093             5.1             2.0  
79302             7.5             3.0  
40023             7.3             4.0  
88936             6.6             2.0  
40999             6.2             4.0  
56672             6.6             4.0

In [37]:
import numpy    #numpy计算包
import jieba    #分词包
import re
#jieba.enable_parallel() # 关闭并行分词
jieba.enable_parallel(8) # 开启并行分词模式，参数为并行进程数 

content = npl_df[npl_df['title']=='寻梦环游记'].comment.values.tolist()
segment = []
for line in content:
    try:
        segs = jieba.lcut(line)
        for seg in segs:
            if len(seg)>1 and seg!='\r\n':
                segment.append(seg)
    except:
        print(line)
        continue
        

words_df=pd.DataFrame({'segment':segment})
#停用词
stopwords=pd.read_csv("stopwords.txt",index_col=False,quoting=3,sep="\t",names=['stopword'], encoding='utf-8')#quoting=3全不引用
words_df=words_df[~words_df.segment.isin(stopwords.stopword)]
#词频统计
words_stat = words_df.groupby('segment')['segment'].agg({'count':'size'}).reset_index().sort_values(by=["count"],ascending=False)

/home/bigdata/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:25: FutureWarning: using a dict on a Series for aggregation
is deprecated and will be removed in a future version
Process ForkPoolWorker-23:
Process ForkPoolWorker-22:
Process ForkPoolWorker-18:
Process ForkPoolWorker-21:
Process ForkPoolWorker-20:
Process ForkPoolWorker-19:
Process ForkPoolWorker-17:
Process ForkPoolWorker-24:
Traceback (most recent call last):
Traceback (most recent call last):
Traceback (most recent call last):
Traceback (most recent call last):
Traceback (most recent call last):
  File "/home/bigdata/anaconda3/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
Traceback (most recent call last):
  File "/home/bigdata/anaconda3/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/home/bigdata/anaconda3/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
Traceback (most recent call last):
  File "/

In [38]:
words_stat.head(10)

segment  count
2796     皮克斯    112
2100      故事     96
2755      电影     96
421       世界     94
540       亡灵     78
229       一个     75
1266     墨西哥     69
2330      梦想     61
2384      死亡     60
3690      音乐     60

#### 重要词抽取,给予tf-idf的抽取

In [39]:
import jieba.analyse as analyse
import pandas as pd
content = "".join(content)
#print(" ".join)
analyse.extract_tags(content, topK=10, withWeight=True, allowPOS=())

[('皮克斯', 0.14757117507336492),
 ('亡灵', 0.06879938469606824),
 ('电影', 0.06058355014230143),
 ('故事', 0.055692507881584835),
 ('动画', 0.05155355050208531),
 ('墨西哥', 0.05008846944738199),
 ('梦想', 0.044810859516145024),
 ('世界', 0.03891448001171374),
 ('音乐', 0.03770427931956398),
 ('套路', 0.03659595226513744)]

In [40]:
import jieba.analyse as analyse
import pandas as pd

(analyse.textrank(content, topK=20, withWeight=True, allowPOS=('ns','n','vn')))


[('故事', 1.0),
 ('电影', 0.8668683339873232),
 ('亡灵', 0.8497028314355037),
 ('墨西哥', 0.789898676471442),
 ('世界', 0.7005461790365071),
 ('音乐', 0.6173455852594754),
 ('家庭', 0.546327163573403),
 ('梦想', 0.523774878509023),
 ('动画', 0.5003970629730443),
 ('文化', 0.41724868281452165),
 ('记忆', 0.3382486600614563),
 ('亲情', 0.3241545739167724),
 ('套路', 0.2997329983698173),
 ('剧情', 0.28458186354115034),
 ('家人', 0.2574312451366974),
 ('时候', 0.24129007242782752),
 ('骷髅', 0.22693022700993812),
 ('情感', 0.2215232310676254),
 ('观众', 0.2197127353815787),
 ('细节', 0.19481810327737428)]

#### 主题模型(LDA)

In [41]:
from gensim import corpora, models, similarities
import gensim
stopwords=pd.read_csv("stopwords.txt",index_col=False,quoting=3,sep="\t",names=['stopword'], encoding='utf-8')#quoting=3全不引用
stopwords.stopword.values

array([',', '?', '、', ..., '上', '上来', '上去'], dtype=object)

In [42]:
import math
def is_sqr(x):
    return math.sqrt(x) % 1 == 0
 
newlist = list(filter(is_sqr, range(1, 101)))
print(newlist)

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]


In [43]:
content = npl_df[(npl_df['title']=='寻梦环游记')][:20].comment.values.tolist()
sentences = []
for line in content:
    try:
        segs = jieba.lcut(line)
        
        segs = list(filter(lambda x:len(x)>1, segs))
        segs = list(filter(lambda x:x not in stopwords.stopword.values, segs))
        sentences.append(segs)
    except:
        print(line)
        continue
        

#### 词袋模型

In [44]:
dictionary = corpora.Dictionary(sentences)
corpus = [dictionary.doc2bow(sentence) for sentence in sentences]
corpus

[[(0, 1),
  (1, 1),
  (2, 1),
  (3, 1),
  (4, 1),
  (5, 1),
  (6, 1),
  (7, 1),
  (8, 1),
  (9, 1),
  (10, 1),
  (11, 1),
  (12, 1),
  (13, 1),
  (14, 1),
  (15, 1),
  (16, 1),
  (17, 1)],
 [(3, 1),
  (18, 1),
  (19, 1),
  (20, 1),
  (21, 1),
  (22, 1),
  (23, 1),
  (24, 1),
  (25, 1),
  (26, 1),
  (27, 1),
  (28, 1),
  (29, 1),
  (30, 1),
  (31, 1),
  (32, 1),
  (33, 1),
  (34, 1),
  (35, 1),
  (36, 1),
  (37, 1),
  (38, 1),
  (39, 1),
  (40, 1),
  (41, 1),
  (42, 1),
  (43, 1),
  (44, 1),
  (45, 1),
  (46, 1),
  (47, 1),
  (48, 1),
  (49, 1),
  (50, 1),
  (51, 1),
  (52, 1),
  (53, 1),
  (54, 1),
  (55, 1),
  (56, 1),
  (57, 1),
  (58, 1),
  (59, 1),
  (60, 1),
  (61, 1)],
 [(3, 1),
  (58, 1),
  (62, 1),
  (63, 2),
  (64, 1),
  (65, 1),
  (66, 1),
  (67, 1),
  (68, 1),
  (69, 1),
  (70, 1),
  (71, 2),
  (72, 1),
  (73, 1),
  (74, 1),
  (75, 1),
  (76, 1),
  (77, 1),
  (78, 1),
  (79, 1),
  (80, 1),
  (81, 1),
  (82, 1)],
 [(80, 1),
  (82, 1),
  (83, 1),
  (84, 1),
  (85, 1),
  (86, 1

#### LDA建模

In [45]:
lda = gensim.models.ldamodel.LdaModel(corpus=corpus, id2word=dictionary, num_topics=20)
print(lda.print_topic(3, topn=5))

0.060*"家人" + 0.021*"逻辑" + 0.021*"成功" + 0.021*"幸好" + 0.021*"为名"


In [46]:
for topic in lda.print_topics(num_topics=10,num_words=10):
    print(topic[1])

0.025*"动画" + 0.018*"亡灵" + 0.018*"皮克斯" + 0.018*"动人" + 0.011*"纷至沓来" + 0.011*"合家欢" + 0.011*"流畅" + 0.011*"之后" + 0.011*"作品" + 0.011*"入关"
0.020*"故事" + 0.011*"编剧" + 0.011*"一波三折" + 0.011*"阴间" + 0.011*"充满" + 0.010*"伯顿" + 0.010*"绚烂" + 0.010*"有没有" + 0.010*"生老病死" + 0.010*"层次分明"
0.021*"故事" + 0.012*"人间" + 0.012*"悲欣" + 0.012*"交集" + 0.012*"俨然" + 0.012*"想象力" + 0.011*"力荐" + 0.011*"新娘" + 0.011*"不敢" + 0.011*"骷髅"
0.060*"家人" + 0.021*"成功" + 0.021*"幸好" + 0.021*"逻辑" + 0.021*"为名" + 0.021*"区别" + 0.021*"以爱" + 0.021*"家长" + 0.021*"人生" + 0.021*"梦想"
0.039*"一边" + 0.039*"情节" + 0.020*"意义" + 0.020*"一个" + 0.020*"故事" + 0.020*"探讨" + 0.020*"推荐" + 0.020*"指导" + 0.020*"皮克斯" + 0.020*"画面"
0.039*"任性" + 0.021*"小孩" + 0.021*"害怕" + 0.021*"情绪" + 0.021*"女人" + 0.021*"原因" + 0.021*"一直" + 0.021*"讲些" + 0.021*"爱恨" + 0.021*"惊喜"
0.029*"元素" + 0.018*"墨西哥" + 0.018*"煽情" + 0.018*"代价" + 0.018*"巧妙" + 0.018*"偶像" + 0.018*"成长" + 0.018*"依旧" + 0.018*"招架" + 0.018*"实在"
0.062*"电影院" + 0.062*"唏嘘" + 0.032*"皮克斯" + 0.032*"孩子" + 0.032*"哭成" + 0.032*"一起" + 0.032*"大人"

## 构建模型

### 将词语分类并记录其位置

In [47]:
# """
# 1. 文本切割
# """

def sent2word(sentence):
    segList = jieba.cut(sentence)
    segResult = []
    for w in segList:
        segResult.append(w)
    with open('stopwords.txt', 'r') as f:
        stopwords = [w.strip() for w in f.readlines()]

    newSent = []
    for word in segResult:
        if word in stopwords:
            # print "stopword: %s" % word
            continue
        else:
            newSent.append(word)
    return newSent
wordDict = sent2word(npl_df[(npl_df['title']=='寻梦环游记')][:20].comment.values[0])

In [48]:
# """
# 2. 情感定位
# """
from collections import defaultdict
import os

def classifyWords(wordDict):
    # (1) 情感词
    with open('BosonNLP_sentiment_score.txt', 'r') as f:
        senList = [w.strip() for w in f.readlines()]

    senDict = {}
    for s in senList:
        try:
            senDict[s.split(' ')[0]] = s.split(' ')[1]
        except:
            continue
    # (2) 否定词
    with open('notDict.txt', 'r') as f:
        notList = [w.strip() for w in f.readlines()]
    # (3) 程度副词
    with open('degreeDict.txt', 'r') as f:
        degreeList = [w.strip() for w in f.readlines()]
    degreeDict = {}
    for s in degreeList:
        try:
            degreeDict[s.split(',')[0]] = s.split(',')[1]
        except:
            continue

    
    senWord = {}
    notWord = {}
    degreeWord = {}
    
    for word in wordDict:
        if word in senDict.keys() and word not in notList and word not in degreeDict.keys():
            senWord[word] = senDict[word]
        elif word in notList and word not in degreeDict.keys():
            notWord[word] = -1
        elif word in degreeDict.keys():
            degreeWord[word] = degreeDict[word]
    return senWord, notWord, degreeWord
senWord, notWord, degreeWord = classifyWords(wordDict)

In [49]:
# """
# 3. 情感聚合
# """
def scoreSent(senWord, notWord, degreeWord, segResult):
    W = 1
    score = 0
    # 存所有情感词的位置的列表
    senLoc = 0
    notLoc = 0
    degreeLoc = 0
    #senloc = -1
    # notloc = -1
    # degreeloc = -1
    
    # 遍历句中所有单词segResult，i为单词绝对位置
    for i in range(0, len(segResult)):
        # 如果该词为情感词
        if segResult[i] in senWord.keys():
            # 直接添加该情感词分数
            score += W * float(senWord[segResult[i]])
            
            # loc为情感词位置列表的序号
            senLoc += 1
            #print("score = %f" % score)
#             if senloc < len(senLoc) - 1:
#                 # 判断该情感词与下一情感词之间是否有否定词或程度副词
#                 # j为绝对位置
#                 for j in range(senLoc[senloc], senLoc[senloc + 1]):
#                     # 如果有否定词
#                     if j in notLoc:
#                         W *= -1
#                     # 如果有程度副词
#                     elif j in degreeLoc:
#                         W *= float(degreeWord[j])
#         # i定位至下一个情感词
#         if senloc < len(senLoc) - 1:
#             i = senLoc[senloc + 1]
    return score

scoreSent(senWord, notWord, degreeWord, wordDict)

6.522944806467

In [50]:
npl_df[npl_df['title']=='寻梦环游记'][:20]

comment  title  \
4571  广州试片会。主持人在现场透露：Coco原本涉及到亡灵题材，是不能引进的。但是在过审时，当场看...  寻梦环游记   
4572  供奉的遗像是牵引家人回家的通道，驻留的记忆是保持亡灵存续的神力，热闹的音乐是唤醒思念启封的药...  寻梦环游记   
4573  不得不服皮克斯，一边和你探讨梦想、成功、家庭的意义；一边用各种充满想象力的情节、画面、音乐，...  寻梦环游记   
4574  对于皮克斯，是很垃圾的一部电影，乏善可陈，没有一点儿新意，看了开头就知道结尾，完全是浪费时间...  寻梦环游记   
4575   Up的一体两面，成长的代价与偶像的崩塌，但墨西哥元素揉的实在太巧妙，煽情起来也依旧让我难以招架。  寻梦环游记   
4576  上映的第四个星期才看，期待很高是真，不如预期也是真。大概是皮克斯正常偏低水准，抓住了“亲情”...  寻梦环游记   
4577                                 可恶的皮克斯，又一次在电影院哭成傻比  寻梦环游记   
4578  11/16@迪斯尼上海 四星半。谢邀试片会，Disney x Pixar这部新作在故事磨合与...  寻梦环游记   
4579  提前宣布这部是2018年奥斯卡最佳动画长片，感恩节之后看简直哭到颤抖，明明是被玩滥的家庭取舍...  寻梦环游记   
4580     请皮克斯收下我的膝盖……看片的时候大家在讨论基佬没有后代，是不是在阴间也会消失得很快。ಥ_ಥ  寻梦环游记   
4582  “Family comes first.”的前提不应该是互相忍让，而是相互理解。（PS.亡灵...  寻梦环游记   
4583  哇，太棒了，阴阳交错，仙境漫游，充满童心的想象力，彩色温暖迷幻，故事一波三折，绚烂的阴间图景...  寻梦环游记   
4584  4.5 在结局猛烈的组合拳煽情下哭到眼睛都睁不开...虽然还是有不少皮克斯惯性的程式，依旧是...  寻梦环游记   
4585               死后世界还是这么三六九等，穷逼死了还是穷逼，那还不如被快点遗忘消失掉啊！  寻梦环游记   
4586                  电影院左右的大人带着孩子一起哭得唏嘘唏嘘的。只是完全不是我的哭点。  寻梦环游记   
4587  人的一生要经历两次死亡：第一次是肉身的死亡，第二次则是从爱你的人的记忆里消失。当这个生者的世...  寻梦环游记   
4588  因为是家人，就有权干涉你的人生、你的梦想，理由是“一家人就要整整齐齐”。片中家人所做的事，跟...  寻梦环游记   
4589  在中段那个可以比肩皮克斯之前作品任何动人瞬间的静默时刻之后，一连串落入窠臼的情节却纷至沓来，...  寻梦环游记   
4590  虽然很不错，但总觉得少了点什么，细想一下原因可能有三：1，祖祖祖奶奶的爱恨表达很草率，唯一的...  寻梦环游记   
4591                            很久没有见过大家纷纷摘下眼镜掏纸巾的动人时刻了  寻梦环游记   

      rating.value_y  rating.value_x  
4571             9.0             5.0  
4572             9.0             5.0  
4573             9.0             5.0  
4574             9.0             2.0  
4575             9.0             5.0  
4576             9.0             3.0  
4577             9.0             5.0  
4578             9.0             4.0  
4579             9.0             5.0  
4580             9.0             5.0  
4582             9.0             5.0  
4583             9.0             5.0  
4584             9.0             5.0  
4585             9.0             4.0  
4586             9.0             3.0  
4587             9.0             4.0  
4588             9.0             3.0  
4589             9.0             3.0  
4590             9.0             4.0  
4591             9.0             4.0

In [51]:
from sklearn.preprocessing import scale
sent2word(npl_df[npl_df['title']=='寻梦环游记'].comment.values[1])
def sentence_to_score(x):
    wordDict = sent2word(x)
    senWord, notWord, degreeWord = classifyWords(wordDict)
    return scoreSent(senWord, notWord, degreeWord, wordDict)
scale(npl_df[npl_df['title']=='寻梦环游记'][:20]['comment'].apply(lambda x:sentence_to_score(x)))

array([-0.40496507,  1.07877657,  0.75825657, -0.88958346, -0.10847657,
        0.79416069, -1.09773176,  1.17644573,  1.83800763, -0.75267598,
       -0.28802841,  1.82640337,  0.85600428, -1.4680098 , -1.09590307,
       -0.8694442 , -0.02236642,  0.47194671, -0.99598886, -0.80682794])

In [52]:
# 这种分数相差很大。。。。感觉用来训练，准确率貌似会很低。。。。

#### 换一种思路，使用Word2Vec构建词义向量

In [53]:
def classify_rating(value):
    if value<=3:
        return 'unlike'
    elif value >= 4:
        return 'like'
    else:
        return 'no_rating'
npl_df['rating_classify'] = npl_df['rating.value_x'].apply(classify_rating)
npl_df['rating_classify'].value_counts()

unlike    52249
like      41176
Name: rating_classify, dtype: int64

In [54]:
#先去掉停用词
def clean_stop_words(comment):
    with open('stopwords.txt', 'r') as f:
        stopwords = set([w.strip() for w in f.readlines()])
    comments=jieba.lcut(comment)
    for word in stopwords:
        while word in comments:
            comments.remove(word)
    return comments
npl_df.loc[:,'comment'] = npl_df.loc[:,'comment'].apply(clean_stop_words)
npl_df.sample(5)

comment  title  \
72600  [BBC, 英伦, 范儿, 很足, 女一, 孔雀蓝, 长裙, 真, 精致, 俄国, 名著, ...  麦克白夫人   
81247       [回望, 人生, 一切都是, 错, 刚刚, 好, 接受, 平庸, 相信, 伟大, 重要]  你好布拉德   
32289       [不忍, 直视, 阴阳, 路, 二十年, 情怀, 蔡卓妍, 演技, 尴尬, 癌, 要犯]  常在你左右   
50962           [12, 岁, 高智商, 熊, 孩子, 恶少, 完成, 变态, 绑架, 杀人案]   安全邻域   
29688  [When,  , I,  , think,  , about,  , that,  , l...    老幸运   

       rating.value_y  rating.value_x rating_classify  
72600             6.5             2.0          unlike  
81247             7.5             3.0          unlike  
32289             5.1             2.0          unlike  
50962             6.4             3.0          unlike  
29688             7.5             4.0            like

为了避免分类学习器偏向的问题，对三类样本做1:1:1采样，组成新的样本。<br>
现在改成2分问题了。

In [55]:
vec_df = pd.concat([npl_df[npl_df['rating_classify'] == 'like']
                   ,npl_df[npl_df['rating_classify'] == 'unlike'].sample(41176)])

vec_df['rating_classify'].value_counts()

unlike    41176
like      41176
Name: rating_classify, dtype: int64

In [56]:
from gensim.models.word2vec import Word2Vec
import numpy as np 
from sklearn.externals import joblib
from sklearn.svm import SVC
from sklearn.preprocessing import scale
import datetime
import sys

In [57]:
# 训练集 测试集的切分
from sklearn.model_selection import train_test_split

def load_data():
    X = vec_df['comment'].values
    y = vec_df['rating_classify'].values
    X_train,X_test,y_train,y_test = train_test_split(X,y,stratify=y,random_state=2018) #random_state随机种子
    
    np.save('y_train.npy',y_train)
    np.save('y_test.npy',y_test)
    
    return X_train,X_test

对每一句话词计算并取平均

In [58]:
def build_sentence_vector(text, size, imdb_w2v):
    vec = np.zeros(size).reshape((1,size))
    count = 0.
    for word in text:
        try:
            vec += imdb_w2v[word].reshape((1,size))
            count += 1.
        except KeyError:
            continue
    if count != 0:
        vec /=count
    return vec

计算词向量

In [59]:
def get_train_vecs(x_train, x_test):
    n_dim = 100
    imdb_w2v = Word2Vec(size=n_dim, min_count=3)
    imdb_w2v.build_vocab(x_train)
    imdb_w2v.train(x_train,total_examples=len(x_train), epochs=10)
    train_vecs = np.concatenate([build_sentence_vector(z, n_dim, imdb_w2v) for z in x_train])
    train_vecs = scale(train_vecs)
    np.save('train_vecs.npy', train_vecs)
    print(train_vecs.shape)
    imdb_w2v.train(x_test,total_examples=len(x_train), epochs=10)
    imdb_w2v.save('model.pkl')
    test_vecs = np.concatenate([build_sentence_vector(z, n_dim, imdb_w2v) for z in x_test])
    test_vecs = scale(test_vecs)
    np.save('test_vecs.npy', test_vecs)
    print(test_vecs.shape)

获取 训练集、测试集、分类标签

In [60]:
def get_data():
    train_vec = np.load('train_vecs.npy')
    y_train = np.load('y_train.npy')
    test_vecs = np.load('test_vecs.npy')
    y_test = np.load('y_test.npy')
    return train_vec,y_train,test_vecs,y_test

#### SVM

In [61]:
def svm_train(train_vecs, y_train, test_vecs, y_test):
    clf = SVC(kernel = 'linear', verbose = False) #'rbf'
    clf.fit(train_vecs, y_train)
    joblib.dump(clf, 'svm_model_1.pkl')
    print(clf.score(test_vecs,y_test))

预测

In [127]:
def get_predict_vecs(words):
    n_dim = 100
    imdb_w2v = Word2Vec.load('model.pkl')
    train_vecs = build_sentence_vector(words, n_dim, imdb_w2v)
    return train_vecs
def predict(string):
    words=jieba.lcut(string)
    words_vec = get_predict_vecs(words)
    clf=joblib.load('svm_model_1.pkl')
    result=clf.predict(words_vec)
    print('%s......  --> svm:  %s'%(s[:10],result))

训练函数

In [63]:
def train():
    print('loading Data......')
    x_train, x_test = load_data()
    get_train_vecs(x_train,x_test)
    print('Model training......')
    train_vecs,y_train,test_vecs,y_test = get_data()
#     svm_train(train_vecs,y_train,test_vecs,y_test)


开始训练模型

In [ ]:
startTime = datetime.datetime.now()
train()
endTime = datetime.datetime.now()
print(endTime-startTime)

**训练结果有点儿惨：[LibSVM]0.413627639155。**
* 做了向量化后，加入一个scale试试。训练了n久，结果：0.43769993602,还是惨
* 只有提高把向量化参数 n_dim = 30，改 n_dim =100了，结果训练了1:20:14.150902，然后提高了一点点0.469449776072
* 有用xgb试了试，也很惨，训练时间5秒钟，correct=0.334693

#### 感觉导致结果比较差的原因是不是因为之前数据处理的时候就有问题。回过去重新检查一下数据
* 提前做了一个停用词的处理，将df中所有的comments处理了停用词，结果：训练了0:39:22.197936，然后降低了一点点0.455534229047

#### 经过寒神的指教，准备把三分问题变为2分问题
* 小于等于3星 dislike, 大于3星 like
* 然后xgb就运作的很好了，svm一直没有跑成功，不知道为何

In [64]:
vec_df.head(1)

comment  title  rating.value_y  \
0  [笑, 太, 疯癫, 笑, 爱, 烂, 片, 影史, 最, 经典, 烂片, 诞生, 迷影, ...  灾难艺术家             7.8   

   rating.value_x rating_classify  
0             5.0            like

#### 用xgb测试一下

In [149]:
import os
import numpy as np
from sklearn.model_selection import GridSearchCV
import xgboost as xgb
os.environ["OMP_NUM_THREADS"] = "8"  #并行训练
rng = np.random.RandomState(2017)
import warnings
warnings.filterwarnings("ignore")


X = vec_df['rating.value_y'].values
y = vec_df['rating_classify'].values
X_train,X_test,y_train,y_test = train_test_split(X,y,stratify=y,random_state=2018) #random_state随机种子


train_vecs,y_train,test_vecs,y_test = get_data()

def get_y_int(y):
    for i in range(0,len(y)):
        if y[i] == 'unlike':
            y[i] = 0
        elif y[i] == 'like':
            y[i] = 1
        else:
            y[i] = 999
    return y

y_train_int = get_y_int(y_train)
y_test_int = get_y_int(y_test)

# X = pd.DataFrame(train_vecs)
# X['score_now'] = X_train
# X = X[[c for c in X.columns]].values

X = train_vecs
y = y_train_int



# param_grid = {
#               'max_depth': [3, 4, 5],
#               'n_estimators': [20, 40],   #, 50, 60, 80, 100, 200, 400
#               'learning_rate': [0.01, 0.02],  #, 0.05, 0.1, 0.2
#               'subsample': [0.65, 0.7, 0.8],
#               'colsample_bylevel':[0.65] #, 0.7, 0.8
#              }

param_grid = {
              'max_depth': [3],
              'n_estimators': [20],
              'learning_rate': [0.01, 0.02],
              'subsample': [0.65],
              'colsample_bylevel':[0.65]
             }



xgb_model = xgb.XGBRegressor(random_state=2017)
rgs = GridSearchCV(xgb_model, param_grid, n_jobs=8, cv=5)
rgs.fit(X, y)
print(rgs.best_score_)
print(rgs.best_params_)

0.0426923440674
{'colsample_bylevel': 0.65, 'learning_rate': 0.02, 'max_depth': 3, 'n_estimators': 20, 'subsample': 0.65}


0.0426923440674<br>
{'colsample_bylevel': 0.65, 'learning_rate': 0.02, 'max_depth': 3, 'n_estimators': 20, 'subsample': 0.65}<br>
0.0866109239447<br>
{'colsample_bylevel': 0.65, 'learning_rate': 0.02, 'max_depth': 5, 'n_estimators': 40, 'subsample': 0.65}

In [138]:
joblib.dump(rgs, 'xgb_model_1.pkl')

['xgb_model_1.pkl']

In [150]:
def xgb_predict(string):
    words=jieba.lcut(string)
    words_vec = get_predict_vecs(words)
    clf=joblib.load('xgb_model_2.pkl')
    result=clf.predict(words_vec)
    if result > 0.5:
        print('%s......  --> xgb:  喜爱'%(s[:10]))
    else:
        print('%s......  --> xgb:  不喜爱'%(s[:10]))

In [151]:
s = '供奉的遗像是牵引家人回家的通道，驻留的记忆是保持亡灵存续的神力，热闹的音乐是唤醒思念启封的药引。我为你写了首歌，穿越浩瀚的岁月烟尘，捱过冰冷的孤独冬季，横跨漫长的天人之路，在你老去的时候，唱给你听。这瑰丽的灯火万家，摇曳的烛光千盏，不如你梳着麻花辫坐在床头时，眼里闪烁的星光璀璨。'
xgb_predict(s)
predict(s)

供奉的遗像是牵引家人......  --> xgb:  喜爱
供奉的遗像是牵引家人......  --> svm:  ['unlike']


In [152]:
s = '无后同性恋们的一场极大的噩梦。'
xgb_predict(s)
predict(s)

无后同性恋们的一场极......  --> xgb:  不喜爱
无后同性恋们的一场极......  --> svm:  ['normal']


In [153]:
s = '漫天好评太夸张了吧。。。Pixar的叙事与价值观内核真是愈发俗套保守了，这还真是个全世界都要回归家庭本位的年代咩。。。'
xgb_predict(s)
predict(s)

漫天好评太夸张了吧。......  --> xgb:  不喜爱
漫天好评太夸张了吧。......  --> svm:  ['unlike']


In [154]:
s = r'最后胡巴跟笨笨跳舞的彩蛋竟然是全场最佳，不断脑补是我GAI跟花园宝宝在唱歌……除此以外剧情大概适合三至五岁小宝宝。'
xgb_predict(s)
predict(s)

最后胡巴跟笨笨跳舞的......  --> xgb:  不喜爱
最后胡巴跟笨笨跳舞的......  --> svm:  ['normal']


In [155]:
s = r'过年看个热闹吧，比第一部差劲'
xgb_predict(s)
predict(s)

过年看个热闹吧，比第......  --> xgb:  不喜爱
过年看个热闹吧，比第......  --> svm:  ['like']


In [156]:
comment = '名著被毁不是第一次。除了交心的《女儿情》，美妆造型都是辣眼睛，梁咏琪国师扮相看到自己也想哭吧？赵丽颖至少演得比冯绍峰好一点。除了特效做得精致，其他东西都牺牲了。奇幻、爱情、搞笑，尽力达到春节档各种元素。重工业大片剧本也不能轻易落下，为了资本投降可违，毕竟好故事总是能打动人的。'
xgb_predict(comment)
predict(comment)

过年看个热闹吧，比第......  --> xgb:  不喜爱
过年看个热闹吧，比第......  --> svm:  ['normal']


In [157]:
comment = '看了点映，女儿国这部真正适合到大屏幕上观看，3D视觉效果极为震撼，超过前作《三打》很多，赵丽颖大银幕上太美了，我要是唐僧一样走不动了。'
xgb_predict(comment)
predict(comment)

过年看个热闹吧，比第......  --> xgb:  喜爱
过年看个热闹吧，比第......  --> svm:  ['like']


In [158]:
comment = '那一瞬你从天而降，这一次我奋身而跃，是心跳的重叠，亦是天命的交错。初遇泪目，别离无奈，再特立独行的爱情也敌不过所谓的使命，什么王权富贵戒律清规，什么爱一人爱众生，什么若离女儿国万物凋零，俱是佛祖的棋子。世俗毁掉你的真心，还要把它解作执念。苦海无涯，命运结界；执子之袖，若有来生。'
xgb_predict(comment)
predict(comment)

过年看个热闹吧，比第......  --> xgb:  喜爱
过年看个热闹吧，比第......  --> svm:  ['unlike']


#### xgb比svm要准确得多